package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,335 @@
+
+        Returns
+        -------
+        self._dim : list
+            Array-like used to slice the positions to obtain desired dimensionality
+        self.dim_fac : float
+            Dimension factor :math:`d` of the MSD.
+
+        """
+        keys = {'x':0, 'y':1, 'z':2}
+        dim = []
+
+        self.msd_type = self.msd_type.lower()
+        print(self.msd_type)
+        try:
+            for tok in self.msd_type:
+                vals = (keys.pop(tok))
+                dim.append(vals)
+        except KeyError:
+            raise ValueError('invalid msd_type specified, please specify one of xyz, xy, xz, yz, x, y, z ')
+        
+        dim.sort()
+        self._dim = dim
+        self.dim_fac = float(len(dim))
+
+    def _select_reference_positions(self):
+        r""" Constructs array of positions for MSD calculation.
+        
+        Parameters
+        ----------
+        self.u : :class:`Universe`
+            MDAnalysis Universe
+        self.selection : str
+            MDAnalysis selection string
+        
+        Returns
+        -------
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape = (N_frames, N_particles, 3)
+        self.N_particles : float
+            Number of particles used in the MSD calculation
+
+        """
+        #avoid memory transfer by iterating 
+        idxs = self.u.select_atoms(self.selection).ix
+        position_array_dummy = np.zeros([len(self.u.trajectory), len(self.u.atoms), 3])
+        for idx,ts in enumerate(self.u.trajectory):
+           position_array_dummy[idx,:,:] = self.u.atoms.positions
+        self._position_array = position_array_dummy[:,idxs, :]
+        self.N_particles = self._position_array.shape[1]
+
+    
+    def run(self):
+        r""" Wrapper to calculate the MSD via either the "simple" or "fft" algorithm.
+        """
+        if self.fft == True:
+            self.timeseries = self._run_fft()
+        else:
+            self.timeseries = self._run_simple()
+
+    def _run_simple(self): # naieve algorithm without FFT
+        r""" Calculates the MSD via the naieve "windowed" algorithm.
+        
+        Parameters
+        ----------
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape (N_frames, N_particles, 3).
+
+        Returns
+        -------
+        self.timeseries : :class:`np.ndarray`
+            The MSD as a function of lag-time.
+
+        """
+        msds_byparticle = np.zeros([self.N_frames, self.N_particles])
+        lagtimes = np.arange(1,self.N_frames)
+        msds_byparticle[0,:] = np.zeros(self.N_particles) # preset the zero lagtime so we dont have to iterate through
+        for n in range(self.N_particles):
+            for lag in lagtimes:
+                disp = self._position_array[:-lag,n,self._dim if lag else None] - self._position_array[lag:,n,self._dim]
```

Comment:
``lag`` will never be falsy because you've defined it between a range of 1 and ``N_frames``

Reply:
You should be able to get rid of the outer loop, I think with something like the below? @tylerjereddy would know more

```
disp = self._position_array[:-lag, :, self._dim] - self._position_array[lag:, :, self._dim]
msds_by_particle[lag] = disp.sum(axis=-1) ** 2
```

Reply:
Looks worth a try--I guess this is a little mind bending because of the 3D array data structure for `_position_array`. If it is already covered by unit tests that will make it easier to try to improve of course.

Reply:
If you do subclass AnalysisBase, you may be able to get rid of the `for lag in lagtimes` loop as well (or more accurately merge it into the loop that `run()` always calls). Not sure how this will work with `tidynamics`.

```python
def _single_frame(self):
  self.disp[:-self.frame_index] += self.ag.positions[self._dim]
  self.disp[self.frame_index:] -= self.ag.positions[self._dim]

def _conclude(self):
  self.msds_by_particle = self.disp**2
```

Reply:
I removed the outer loop over n_particles.

Reply:
Moving the work into the run loop I think is a bit suboptimal, as the data still needs to be accumulated for the tidynamics FFT based option. Happy to discuss.